### ISSUE TITLE
Unsound: able to drop the message while some objects still point into it

### ISSUE THREAD

**Issue by Hawk777** [herblet/stomp-parser#7]
Consider this code:
```rust
use stomp_parser::client::ClientFrame;
use stomp_parser::headers::DestinationValue;

fn foo() -> DestinationValue {
	let message = b"SEND\n\
                destination:stairway/to/heaven\n\
                \n\
                Lorem ipsum dolor sit amet,...\x00"
                .to_vec();
	let frame = ClientFrame::try_from(message).unwrap();
	let ClientFrame::Send(frame) = frame else { panic!(); };
	frame.destination
}

fn main() {
	let dest_header = foo();
}
```

This does not contain the word `unsafe` anywhere; however, it fails `cargo miri run`:
```
error: Undefined Behavior: constructing invalid value at .value: encountered a dangling reference (use-after-free)
  --> src/main.rs:16:20
   |
16 |     let dest_header = foo();
   |                       ^^^^^ constructing invalid value at .value: encountered a dangling reference (use-after-free)
   |
   = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
   = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
   = note: BACKTRACE:
   = note: inside `main` at src/main.rs:16:20: 16:25
   = note: inside `<fn() as std::ops::FnOnce<()>>::call_once - shim(fn())` at /tmp/rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:250:5: 250:71
   = note: inside `std::sys::backtrace::__rust_begin_short_backtrace::<fn(), ()>` at /tmp/rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys/backtrace.rs:154:18: 154:21
   = note: inside closure at /tmp/rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/rt.rs:195:18: 195:75
   = note: inside `std::ops::function::impls::<impl std::ops::FnOnce<()> for &dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe>::call_once` at /tmp/rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:284:13: 284:31
   = note: inside `std::panicking::r#try::do_call::<&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe, i32>` at /tmp/rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:557:40: 557:43
   = note: inside `std::panicking::r#try::<i32, &dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe>` at /tmp/rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:520:19: 520:88
   = note: inside `std::panic::catch_unwind::<&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe, i32>` at /tmp/rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panic.rs:358:14: 358:33
   = note: inside closure at /tmp/rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/rt.rs:174:48: 174:73
   = note: inside `std::panicking::r#try::do_call::<{closure@std::rt::lang_start_internal::{closure#1}}, isize>` at /tmp/rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:557:40: 557:43
   = note: inside `std::panicking::r#try::<isize, {closure@std::rt::lang_start_internal::{closure#1}}>` at /tmp/rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:520:19: 520:88
   = note: inside `std::panic::catch_unwind::<{closure@std::rt::lang_start_internal::{closure#1}}, isize>` at /tmp/rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panic.rs:358:14: 358:33
   = note: inside `std::rt::lang_start_internal` at /tmp/rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/rt.rs:174:20: 174:98
   = note: inside `std::rt::lang_start::<()>` at /tmp/rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/rt.rs:194:17: 199:6
```

I wasn’t able to actually trace what values are stored in which structs inside `stomp-parser` because of all the macros, but I assume that when `message` (the `Vec<u8>`) is moved into `ClientFrame::try_from`, it’s stored inside the resulting `ClientFrame` object, and when that object is dropped, so is the `Vec<u8>`, despite the fact that the `DestinationValue` object has been moved out of the `ClientFrame` object and still contains an `&'static str` pointing into the `Vec<u8>` (something the compiler doesn’t catch for you because you transmuted to gain the `'static` lifetime).

**Comment by herblet** [comment]
Hello Hawk777,
I will look into this ASAP, but I might not get a chance until tomorrow.

It's been a while since I implemented it but I wouldn't be surprised if there is such a problem, it works in my use case (https://github.com/herblet/little-stomper) but by the look of it isn't safe.

**Comment by herblet** [comment]
Hello Hawk777,
working on this (solution seems to work, need to test in a real app).

To answer your question: yes, 'recently' (in commit-count terms) I change the parser to use a zero-copy (or at least minimal-copy) approach, meaning the raw data is kept in the frame and header values are slices in that raw data. I did not correctly change the header values.

In the new version the headers are private and only references can be obtained publicly; the lifetime of those references and their values is tied to the Frame, thus your example would no longer build, since the destination can't be returned after the frame is dropped.

https://github.com/herblet/stomp-parser/pull/8

**Comment by herblet** [comment]
Just merged the PR, and published version 0.7 of the crate.

The header are now private and there are eponymous accessor methods which return a refererence to the header, the lifetime of which is bound by the lifetime of the frame. Similarly, the lifetime of the headers values is bound by the lifetime of the frame. So you either need to keep the frame to access the values, or clone the values.

**Comment by Hawk777** [comment]
Great, I think this seems like a sensible approach, thanks.
